WordPress/Classes/Services/MenusService.m

```diff
@@ -244,6 +200,79 @@ - (void)generateDefaultMenuItemsForBlog:(Blog *)blog
 }
 
+#pragma mark - private
+
+- (void)createMenuWithName:(NSString *)menuName
+                      blog:(Blog *)blog
+                   success:(nullable void(^)(NSNumber *menuID))success
+                   failure:(nullable MenusServiceFailureBlock)failure
+{
+    NSParameterAssert([blog isKindOfClass:[Blog class]]);
+    NSAssert([self blogSupportsMenusCustomization:blog], @"Do not call this method on unsupported blogs, check with blogSupportsMenusCustomization first.");
+    
+    if (blog.wordPressComRestApi == nil) {
+        return;
+    }
+    
+    MenusServiceRemote *remote = [[MenusServiceRemote alloc] initWithWordPressComRestApi:blog.wordPressComRestApi];
+    [remote createMenuWithName:menuName
+                          blog:blog
+                       success:^(RemoteMenu * _Nonnull remoteMenu) {
+                           [self.managedObjectContext performBlock:^{
+                               if (success) {
+                                   success(remoteMenu.menuID);
+                               }
+                           }];
+                       }
+                       failure:failure];
+}
+
+- (void)updateMenu:(Menu *)menu
+           forBlog:(Blog *)blog
+           success:(nullable void(^)())success
+           failure:(nullable MenusServiceFailureBlock)failure
+{
+    NSMutableArray *locationNames = nil;
+    if (menu.locations.count) {
+        locationNames = [NSMutableArray arrayWithCapacity:menu.locations.count];
+        for (MenuLocation *location in menu.locations) {
+            if (location.name.length) {
+                [locationNames addObject:location.name];
+            }
+        }
+    }
+    
+    NSArray *remoteItems = nil;
+    if (menu.items.count) {
+        remoteItems = [self remoteItemsFromMenuItems:menu.items];
+    }
+    
+    MenusServiceRemote *remote = [[MenusServiceRemote alloc] initWithWordPressComRestApi:blog.wordPressComRestApi];
+    [remote updateMenuForID:menu.menuID
+                       blog:blog
+                   withName:menu.name
+              withLocations:locationNames
+                  withItems:remoteItems
+                    success:^(RemoteMenu * _Nonnull remoteMenu) {
+                        [self.managedObjectContext performBlock:^{
+                            /*
+                             Update the local menu with the fresh MenuItems from remote.
+                             We need to replace the MenuItems as it's difficult to keep track of
+                             which items are equal to one another, especially when a menuID is unknown.
+                             */
+                            menu.items = nil;
+                            for (RemoteMenuItem *remoteItem in remoteMenu.items) {
+                                [self addMenuItemFromRemoteMenuItem:remoteItem forMenu:menu];
+                            }
+                            [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
```

Comment:
Maybe `[[ContextManager sharedInstance] saveContext: withCompletionBlock:]` and call success from with in the completion block.  That way there is no race condition between the core data save and whatever is performed in `success`